Scannerlock

daschuer opened the pull request:
I experience a hard to reproduce deadlock in LibraryScanner destructor by starting Mixxx and closing it just after 

This fixes some race condition in that code part and hopefully fixes the deadlock. 

Before this patch It could happen that a scan request is on the LibraryScanner event queue when the destructor is called. 
In this state cancel() is not called. But we will have a running scan later.

daschuer committed:
remove redundant check for m_scannerGlobal not null

daschuer committed:
fix race condition in LibraryScanner::slotCancel()

daschuer committed:
replaced setShouldCancel(true) with just cancel()

daschuer committed:
Add a mutex, to lock the time between the scan request an initated scanner.

daschuer committed:
fix typo

daschuer commented:
Who likes to review this?

illuusio commented:
Is there a way to create this deadlock?

sblaisot commented on src/library/scanner/libraryscanner.cpp:
```diff
@@ -99,11 +99,11 @@ LibraryScanner::LibraryScanner(QWidget* pParentWidget, TrackCollection* collecti
-        // Cancel any running library scan.
-        cancel();
-    }
+    // Wait until there is no pending scan request
```
With this logic, don't we delay Mixxx closing too much ? What is the typical mutex lock time ?

illuusio replied:
Mutex doesn't just disappear they are eternal if you don't release them.

sblaisot replied:
Well, I know. 
My question is: what is the typical time this particular mutex is on (time of a pending scan request if I understand correctly the code).
We wait here for the mutex to be release elsewhere in the code (when there is no more pending request), so how long could this be ? if it's a matter of milliseconds, that's ok, but if there can be conditions where the request can be pending for seconds or minutes, that would delay Mixxx closing in a way that is disturbing to users.

daschuer commented:
> Is there a way to create this deadlock?

I can only create this deadlock by luck on my slow eeePC if you open Mixxx and close it just after. 

Mixxx will normally not wait at all at the mutex at all, except for the rare case you start a scan and shut down mixxx simultaneously. In this case Mixxx waits at the mutex until the scanner is fully established which is a question of milliseconds.

I am currently working on the comments issued by @uklotzde. Thanks to this I think I have found an related issue that it is possible that the Mixxx shoutdown is actually delayed for the scanning time of one folder. In this case it waits at m_pool.waitForDone(). 

> Mutex doesn't just disappear they are eternal if you don't release them. 

The QMutexLocker handles the releasing after the destructor is finished and the locker scope ends.

daschuer committed:
using a state enum to clarify the scanner transitions

daschuer commented:
@uklotzde: now the state variable is in place. I have also tried to clarify the cancel slots a bit. 


There is still the pending issue that a cancel command and shutting down of Mixxx can be delayed significant by a running scan. There we should tweaked the compromise between saving the scanner progress to database and a quick shutdown. 

 @all:
What do you think, should we allow a quick restart of Mixxx in case of emergency, on the expense to loose the scanner progress? Is it required to distinguish between a graceful cancel during a Mixxx run and a quit command during Mixxx shutdown.

illuusio commented:
I vote no emergency because you leave database in unknown conditions. If this only adds couple millis we should make good shut down. In other hand I'm middle of kick and I can't close Mixxx. Hmm... still less headache with no emergency exit or ask from user and let him decide.

daschuer commented:
> I vote no emergency because you leave database in unknown conditions. 

Of cause that is a no go. Even if we decide to have an emergency exit, we must not mess around with the database. It is more question of finish a pending task or break it as it where never started and probably loose the collected data in RAM since than.

kain88-de commented:
So far we always tried to finish it. We should adopt the same strategy here to be consistent.

daschuer committed:
Fix signal Name

daschuer commented:
Ok, I am finished. 
I have added the emergency exit topic to a new PR https://github.com/mixxxdj/mixxx/pull/672

Who likes to review?

daschuer committed:
Added setScannerState to control the state in a single place

daschuer committed:
Added asserts  to setScannerState

daschuer committed:
access m_state and m_stateMutex only in setScannerState

daschuer committed:
renamed setScannerState to changeScannerState since it does mor than just set

daschuer commented:
I am not sure if I have understand how your proposed solution will finally look like,
In the latest commits the mutex and the state are only accessed in the changeScannerState() function. 

We need to lock the Mutex during whole STARTING and CANCELING, since we cannot wait at m_scannerState.

I have also added some more comments and debug asserts.
Better now?

daschuer committed:
be sure mutex is locked when modifying scanner state

daschuer commented:
Now I have changed the code in a way that the mutex is looked before modifying m_state. 
I think it actually fixes a race condition if you cancel just at the time the scanner finishes. 

However the mutex does still not protect m_state write when setting to idle, it still just protects the starting and stopping phase.
There is a possible remaining race if we set idle just before setting canceling but after locking the mutex. I am working for a fix.

daschuer committed:
Added FINISHED Scanner state to avaoid a race condition

daschuer commented:
@uklotzde: I think it should be rock solid now. Please have a look.

uklotzde commented:
LGTM

The code of the scanner tasks is not easy to understand with all the signalling, global state, ... IMHO this could be improved by revising the design and straightening the program flow. I already have some ideas, but that can wait ;)

ywwg commented:
This is a lot of code for a bug-fix. Please please please add tests if you're going to write so much!

Be-ing commented:
@uklotzde : I have ideas for reworking scanning and analysis and how to present them in the GUI. See https://blueprints.launchpad.net/mixxx/+spec/library-scan-analyze-reorganization . Let's talk more after 1.12 is released.

ywwg commented on src/library/scanner/libraryscanner.cpp:
```diff
@@ -170,6 +156,17 @@ void LibraryScanner::run() {
 
 void LibraryScanner::slotStartScan() {
     qDebug() << "LibraryScanner::slotStartScan";
+    DEBUG_ASSERT(m_state == STARTING);
```
please use debug_assert_and_handle so we don't crash.  maybe just return if this invariant doesn't hold

daschuer replied:
If we hit that we have an compile time error, there is nothing to handle. Returning here in non debug will make the things worse.

ywwg replied:
how is the value of m_state a compile-time error?

daschuer replied:
Yes, compile-time error is the wrong word. 
You can only change the value of m_state here, by recompiling Mixxx.

ywwg commented on src/library/scanner/libraryscanner.cpp:
```diff
@@ -347,27 +338,65 @@ void LibraryScanner::slotFinishScan() {
-    if (m_scannerGlobal) {
-        m_scannerGlobal->setShouldCancel(true);
+    // The m_stateMutex is locked
+    DEBUG_ASSERT(m_state == CANCELING);
```
also handle this case

daschuer replied:
The same here. The assert guards only against wrong changes of the code.

ywwg commented on src/library/scanner/libraryscanner.cpp:
```diff
@@ -452,6 +482,66 @@ void LibraryScanner::addNewTrack(TrackPointer pTrack) {
+}
+
+bool LibraryScanner::changeScannerState(ScannerState newState) {
+    // Allowed State transitions:
```
put this documentation in the header

ywwg commented on src/library/scanner/libraryscanner.cpp:
```diff
@@ -452,6 +482,66 @@ void LibraryScanner::addNewTrack(TrackPointer pTrack) {
+            return false;
+        }
+    case SCANNING:
+        DEBUG_ASSERT(m_state == STARTING);
```
handle all debug asserts

daschuer replied:
The same here, nothing to handle. Look at this like unit-tests against later regressions due to misuse of this function. 
I am currently working on dedicated unit tests as well. 
The only remaining issue is that I can't setup a unittest against various race conditions.

uklotzde replied:
I'm with Daniel regarding the assert. We could safely omit them. The asserts are just there for documentation purposes with the bonus of runtime checks during development/testing. Please remember to only "assert for the impossible". All possible errors have to be handled explicitly. But if one of these asserts ever fails we have reached an undefined program state resulting in undefined behavior.

ywwg replied:
> Look at this like unit-tests

runtime checks are not at all like unit tests.  I understand how this could catch regressions, but it's the least-desirable way to test for regressions.

ywwg commented:
Can you please describe a series of steps for testing this patch, like "start up, wait for scan to start, do XXX when you see YYY".

daschuer committed:
Added skeleton for library scanner test

daschuer committed:
Added tests for all transitions to have a 100 % coverage of changeScannerState()

daschuer commented:
> Can you please describe a series of steps for testing this patch, like "start up, wait for scan to start, do XXX when you see YYY".

Now a unit test for the normal transitions is in place. 

The original issue was that the library scanner was not stopped if you start Mixxx and stop it immediately. You may test this, but I was only able to reproduce this on my old netbook in rare cases, so positive manual test results might be not that significant. 

During the development I have introduced and removed an issue that happens if you start and cancel the scanner "at once". I was not able to reproduce this by enabling CPU burning loops. 
So I am afraid you cannot test it manually now.

If you like to spend some time in reviewing this PR, I think it is best to check if there is still a possible race and if this is maintainable. 
I hope it is after introducing @uklotzde's state variable idea, which helps to make the transitions unit-testable, but also made this PR that big.

daschuer committed:
Merge remote-tracking branch 'upstream/1.12' into scannerlock

Conflicts:
	src/library/scanner/libraryscanner.cpp
	src/library/scanner/libraryscanner.h

daschuer committed:
moved comments to header.

daschuer commented:
Notes addressed.

ywwg commented on src/library/scanner/libraryscanner.cpp:
```diff
@@ -69,14 +72,16 @@ LibraryScanner::LibraryScanner(QWidget* pParentWidget, TrackCollection* collecti
-            &(collection->getTrackDAO()), SLOT(databaseTracksMoved(QSet<int>, QSet<int>)));
-    connect(this, SIGNAL(tracksChanged(QSet<int>)),
-            &(collection->getTrackDAO()), SLOT(databaseTracksChanged(QSet<int>)));
+    if (collection) { // false only during test
```
(collection != NULL)  for clarity

ywwg commented on src/library/scanner/libraryscanner.cpp:
```diff
@@ -135,32 +125,33 @@ LibraryScanner::~LibraryScanner() {
-        if (!m_database.open()) {
-            qDebug() << "Failed to open database from library scanner thread." << m_database.lastError();
-            return;
+    if (m_pCollection) { // false only during tests
```
same

ywwg commented on src/library/scanner/libraryscanner.cpp:
```diff
@@ -476,6 +510,58 @@ void LibraryScanner::addNewTrack(TrackPointer pTrack) {
+    case CANCELING:
+        // canceling is always possible, but wait
+        // until there is no scan starting.
+        m_stateMutex.lock();
```
where is this unlocked?

ywwg replied:
It seems that the lock is being used more like a boolean than a thread lock.  The pattern of locking and unlocking in different functions seems to risk deadlocks.  What if CANCEL is fired twice?  Won't it double lock and hang?

uklotzde replied:
The unconventional usage of the mutex was (and still is) my concern. I don't 
remember that I ever used such a pattern when implementing worker threads or 
a thread pool. On the other hand I couldn't find any transition that locks 
or unlocks the mutex twice.

There is a small chance that the mutex does not get unlocked if 
slotStartScan() is left through an exception, because the the mutex is 
unlocked at the end of this function when the state is finally changed to 
SCANNING. @daschuer Would it be possible to change the state earlier to 
SCANNING or IDLE depending on dirs.isEmpty() at the beginning of this function?

On 08/17/2015 07:14 PM, Owen Williams wrote:
>
> In src/library/scanner/libraryscanner.cpp 
> <https://github.com/mixxxdj/mixxx/pull/655#discussion_r37211353>:
>
> > +            return true;
> > +        } else {
> > +            qDebug() << "LibraryScanner: mutex locked, state =" << m_state;
> > +            return false;
> > +        }
> > +    case SCANNING:
> > +        DEBUG_ASSERT(m_state == STARTING);
> > +        // Transition protected by the mutex is over now
> > +        // Allow canceling
> > +        m_state = SCANNING;
> > +        m_stateMutex.unlock();
> > +        return true;
> > +    case CANCELING:
> > +        // canceling is always possible, but wait
> > +        // until there is no scan starting.
> > +        m_stateMutex.lock();
>
> It seems that the lock is being used more like a boolean than a thread 
> lock. The pattern of locking and unlocking in different functions seems to 
> risk deadlocks. What if CANCEL is fired twice? Won't it double lock and hang?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/655/files#r37211353>.
>

daschuer replied:
> where is this unlocked?
It is unlocked by changeScannerState(IDLE);
I will add a comment.

daschuer replied:
> What if CANCEL is fired twice? Won't it double lock and hang?

CANCEL and IDLE are set from the same thread, in a single call, so there is no chance for a deadlock. 

Normally semaphores are used for inter thread communication. We have no need for the counting nature of semaphores here. Should we switch to a semaphore anyway, to clarify that?

daschuer replied:
> Would it be possible to change the state earlier to SCANNING or IDLE depending on dirs.isEmpty() at the beginning of this function?

Good point. Yes this should work since we don't depend on the value of m_scannerGlobal any more.

ywwg replied:
yes I'd prefer a semaphore

ywwg commented on src/library/scanner/libraryscanner.h:
```diff
@@ -83,14 +90,35 @@ class LibraryScanner : public QThread {
+    void cancelAndQuit();
+    void cancel();
+
+    // Allowed State transitions:
```
what happens if a state transition happens that is not allowed?

daschuer replied:
I have tried to consider all possible transactions and catch the not allowed by assertions. 
Do I miss one?

ywwg commented:
Thanks for the tests.  I still have concerns about edge cases of strange state transition requests, and I think we need to protect against those.

daschuer commented:
> Thanks for the tests. I still have concerns about edge cases of strange state transition requests, and I think we need to protect against those.

Which one?

daschuer committed:
Some more comments

daschuer committed:
moved changeScannerState(SCANNING) to the top of slotStartScan()

daschuer commented:
Notes addressed.

ywwg commented:
> Which one?

my biggest concern is a double-cancel.  I can imagine some strange input mechanism firing that slot twice.  Any user-initiated function call is prone to strange behavior and should be treated as guilty until proven innocent.

daschuer committed:
Switch from a Mutex to a Semaphore to protect the Scanner transactions

daschuer committed:
update coments to the semaphore nomenclature

daschuer committed:
Added assertion to guard against canceling twice

daschuer commented:
Now we have a Semaphore. 

I have also added an assertion against canceling twice. 
But this cannot happend with the current code since CANCELING is always terminated with IDLE in the same calling function.

ywwg commented on src/library/scanner/libraryscanner.cpp:
```diff
@@ -69,14 +73,16 @@ LibraryScanner::LibraryScanner(QWidget* pParentWidget, TrackCollection* collecti
-            &(collection->getTrackDAO()), SLOT(databaseTracksChanged(QSet<int>)));
+    if (collection != NULL) { // false only during test
+        connect(this, SIGNAL(scanFinished()),
+                &(collection->getTrackDAO()), SLOT(clearCache()));
```
nit for clarity: pull out collection->getTrackDAO() into a variable

ywwg commented:
one last note (see above), then LGTM

ywwg commented:
I'll go ahead and do the nit fix since it's old code anyway
